tensorboard/loader.py

```diff
@@ -732,85 +740,367 @@ def get_next_event(self):
+    :rtype: tf.Event
+    """
+    self.mark()
+    result = self.get_next_event()
+    self.reset()
+    return result
+
+  def get_offset(self):
+    """Returns number of bytes read across all event log files.
 
     :rtype: int
     """
     if self._mark != -1:
       return self._mark
-    return self._offset
+    return self._get_offset()
+
+  def _get_offset(self):
+    return sum(el.get_offset() for el in self._logs) + self._entombed_progress
 
   def get_size(self):
-    """Returns byte length of file.
+    """Returns sum of byte lengths of event log files.
 
     :rtype: int
     """
-    return self._reader.get_size()
+    return sum(el.get_size() for el in self._logs) + self._entombed_progress
+
+  def save_progress(self):
+    """Saves current offsets of all open event logs to DB.
+
+    This should be called after the mark has been advanced.
+    """
+    with contextlib.closing(self._db_conn.cursor()) as c:
+      n = 0
+      while self._i >= n < len(self._logs):
+        log = self._logs[n]
+        n += 1
+        offset = log.get_offset()
+        c.execute(
+            'UPDATE EventLogs SET offset = ? WHERE rowid = ? AND offset < ?',
+            (offset, log.rowid, offset))
 
   def mark(self):
     """Marks current position in file so reset() can be called."""
     if self._prepended_events:
       raise ValueError('mark() offsets must be monotonic')
-    self._mark = self._offset
+    self._mark = self._get_offset()
     self._saved_events.clear()
 
   def reset(self):
     """Resets read state to where mark() was called."""
     if self._mark == -1:
-      raise ValueError('mark() was not called')
+      return
     self._prepended_events.extend(self._saved_events)
     self._saved_events.clear()
 
   def close(self):
-    """Closes event log reader if open.
+    """Closes all event log readers.
 
-    Further i/o is not permitted after this method is called.
+    This method may be called multiple times, but further operations
+    are not permitted.
+
+    Raises:
+      Exception: To propagate the most recent exception thrown by the
+          EventLogReader close method. Suppressed exceptions are
+          logged.
     """
-    if self._reader is not None:
-      self._reader.close()
-      self._reader = None
+    util.close_all(self._logs)
+    self._i = len(self._logs)
+    self._mark = -1
+    self._prepended_events.clear()
+    self._saved_events.clear()
+
+  def _cleanup(self):
+    # Last event log has to be preserved so we can continue enforcing
+    # monotonicity. We entomb offset because that also has to be
+    # monotonic, but the size does not.
+    if 0 < self._i < len(self._logs):
+      deleted = self._logs[:self._i]
+      self._logs = self._logs[self._i:]
+      self._i = 0
+      self._entombed_progress += sum(l.get_offset() for l in deleted)
+      util.close_all(deleted)
+
+  def _fast_forward_over_things_we_have_read(self):
+    for i in range(len(self._logs) - 1, self._i + 1, -1):
```

Comment:
No longer applies. I don't feel strongly about it but will remember for the future.